rasa/core/featurizers/single_state_featurizer.py

```diff
@@ -40,127 +102,227 @@ class SingleStateFeaturizer:
-        self.entity_tag_specs = self._create_entity_tag_specs(bilou_tagging)
-
-    def _state_features_for_attribute(
-        self, sub_state: SubState, attribute: Text
-    ) -> Dict[Text, int]:
-        # FIXME: the code below is not type-safe, but fixing it
-        #        would require more refactoring, for instance using
-        #        data classes in our states
-        if attribute in {INTENT, ACTION_NAME}:
-            return {sub_state[attribute]: 1}  # type: ignore[dict-item]
-        elif attribute == ENTITIES:
-            return {entity: 1 for entity in sub_state.get(ENTITIES, [])}
-        elif attribute == ACTIVE_LOOP:
-            return {sub_state["name"]: 1}  # type: ignore[dict-item]
-        elif attribute == SLOTS:
-            return {
-                f"{slot_name}_{i}": value
-                for slot_name, slot_as_feature in sub_state.items()
-                for i, value in enumerate(slot_as_feature)
-            }
-        else:
-            raise ValueError(
-                f"Given attribute '{attribute}' is not supported. "
-                f"It must be one of '{self._default_feature_states.keys()}'."
-            )
+        attribute_to_features: Dict[Text, List[Features]] = dict()
+        for attribute in attributes:
+            all_features = parsed_message.get_sparse_features(
+                attribute
+            ) + parsed_message.get_dense_features(attribute)
 
-    def _create_features(
+            for features in all_features:
+                if features is not None:
+                    attribute_to_features[attribute].append(features)
+
+        attribute_list = keep_only_sparse_seq_converted_to_sent or []
+        for attribute in attribute_list:
+            if attribute_to_features.get(attribute):
+                converted_features = convert_sparse_sequence_to_sentence_features(
+                    attribute_to_features[attribute]
+                )
+                if converted_features:
+                    attribute_to_features[attribute] = converted_features
+
+        attribute_list = fallback_to_custom_encoding or []
+        for attribute in fallback_to_custom_encoding:
+            if attribute not in attribute_to_features:
+                attribute_to_features[attribute] = self.featurize_substate_attribute(
+                    sub_state, attribute, sparse=sparse,
+                )
+        return attribute_to_features
+
+    def _featurize_attribute_via_custom_encoding(
         self, sub_state: SubState, attribute: Text, sparse: bool = False
     ) -> List["Features"]:
-        state_features = self._state_features_for_attribute(sub_state, attribute)
+        """Creates a multi-hot encoding like feature for the given attribute.
+
+        Leverages the item2index mappings computed during `prepare_training`.
+
+        FIXME/TODO: would've expected a one-hot encoding type of thing, but for
+          SLOTS the values are >1 ...(?!?)
```

Comment:
Slots can have custom encodings that depend the user's implementation. See [this](https://rasa.com/docs/rasa/domain#custom-slot-types).

Reply:
Oh I see, and you do not know those values in advance and hence cannot map them to categorical "slot_value" type of things. I think it would be good to then clearly point this out somewhere (for us who build algorithms with this and might not expect that they need to scale the values and can squeeze them into some combination layer nonetheless :/).